### PR TITLE
BuildCheck does not run on restore

### PR DESCRIPTION
**Pull request by maridematte**
Fixes: https://github.com/dotnet/msbuild/issues/9747

### Context
We currently run BuildCheck during the restore phase of the build, because of this we end up running BuildCheck twice per project. This PR disables BuildCheck during restore phase.

### Changes Made
Added two check for the restore phase on the code.

The first one is on the `RequestBuilder`, it reads the global property `IsRestore` and creates a BuildCheck instance as necessary.

The second one is within the `BuildCheckConnectorLogger` so we do not start a new BuildCheck run because of received events. In this case I added a ` bool ` to  ` ProjectEvaluationStartedEventArgs` so we can have access to the ` IsRestore` variable within the logger. We skip all event messages until we get a ` ProjectFinishedEventArgs`, which signals the end of a build / the restore phase.

### Tests
Added extra tests that only runs the restore target.

Reopened from https://github.com/dotnet/msbuild/pull/9907

### CODE REVIEW

File: eng/BootStrapMSBuild.props
Comment (JanKrivanek) [inline eng/BootStrapMSBuild.props:1]: This looks like a merge issue - I suspect git casing issue on windows (the file was renamed to BootStrapMsBuild.props - lower 's' - in PR merged to main).

tl;dr; just delete this

File: src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs [hunk 1]
```diff
@@ -47,16 +58,41 @@ private void EventSource_AnyEventRaised(object sender, BuildEventArgs e)
...
+                buildCheckManager.ProcessEvaluationFinishedEventArgs(
+                    loggingContextFactory.CreateLoggingContext(e.BuildEventContext!),
+                    projectEvaluationFinishedEventArgs);
+            }
+            catch (Exception exception)
+            {
+                Debugger.Launch();
+                Console.WriteLine(exception);
+                throw;
+            }
```
Comment (JanKrivanek) [inline src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs:89]: This seems to be merge leftover - this was changed on main

File: src/Build/BuildCheck/Logging/BuildAnalysisLoggingContextExtensions.cs
Comment (JanKrivanek) [inline src/Build/BuildCheck/Logging/BuildAnalysisLoggingContextExtensions.cs:1]: Merge leftover - can be removed

File: src/Build/BuildCheck/Logging/IBuildAnalysisLoggingContext.cs
Comment (JanKrivanek) [inline src/Build/BuildCheck/Logging/IBuildAnalysisLoggingContext.cs:1]: Merge leftover - can be removed

File: src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs [hunk 2]
```diff
@@ -60,12 +68,30 @@ private void HandleProjectEvaluationFinishedEvent(ProjectEvaluationFinishedEvent
...
         {
             _buildCheckManager.StartProjectEvaluation(BuildCheckDataSource.EventArgs, eventArgs.BuildEventContext!, eventArgs.ProjectFile!);
         }
     }
 
+    private void HandleProjectFinishedEvent(ProjectFinishedEventArgs projectFinishedEventArgs)
+    {
+        if (isRestore)
+        {
+            isRestore = false;
```
Comment (rainersigwald) [inline src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs:87]: This doesn't make sense to me, won't there be a bunch of ProjectFinished events that will fire one after the other, causing the first one to unset this and the others to do check stuff?

File: src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs [hunk 3]
```diff
@@ -1105,9 +1105,9 @@ private async Task<BuildResult> BuildProject()
             ErrorUtilities.VerifyThrow(_targetBuilder != null, "Target builder is null");
 
             // We consider this the entrypoint for the project build for purposes of BuildCheck processing 
-
-            var buildCheckManager = (_componentHost.GetComponent(BuildComponentType.BuildCheckManagerProvider) as IBuildCheckManagerProvider)!.Instance;
-            buildCheckManager.SetDataSource(BuildCheckDataSource.BuildExecution);
+            var propertyEntry = _requestEntry.RequestConfiguration.GlobalProperties[MSBuildConstants.MSBuildIsRestoring];
```
[the remark this answers is not in the document]
Comment (maridematte) [reply src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs:1108]: There are two points in BuildCheck that actually run the analyzer. One of them is through the `RequestBuilder` which works on the worker nodes. And through the `BuildCheckConnectorLogger` which runs on the main node. The modification to the evaluation-started event is to get information about the restore phase to the connector logger.

File: src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs (hunk 2 above)
Comment (maridematte) [reply src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs:87]: Yeah, I figured that out after I ran this for a bit. I had hoped the event was just fired once. Is there an event that is fired at the end of the restore phase that I can use to to control the flow on the connector logger?

Comment (JanKrivanek) [reply src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs:87]: If we are relying on the restore happen fully before the build - then the first `ProjectEvaluationStartedEventArgs` that is not marked with `IsRestore` denotes the start of the build after the restore.

Comment (maridematte) [reply src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs:78]: I changed the `isRestore`  reset to happen when we receive another `ProjectEvaluationStartedEventArgs` event.

File: src/Build/Evaluation/Evaluator.cs [hunk 4]
```diff
@@ -626,7 +626,7 @@ private void Evaluate()
                     }
                 }
 
-                _evaluationLoggingContext.LogProjectEvaluationStarted();
+                _evaluationLoggingContext.LogProjectEvaluationStarted(_data.GlobalPropertiesDictionary[MSBuildConstants.MSBuildIsRestoring] is not null); ;
```
Comment (ladipro) [inline src/Build/Evaluation/Evaluator.cs:629]: We currently set the property only to `true` but it's not unconceivable that at some point someone would want to set it to `false` to mean "we're not restoring". For future proofing, would you consider parsing the value instead of just checking that the prop is defined?

File: src/Framework/ProjectEvaluationStartedEventArgs.cs [hunk 5]
```diff
@@ -30,5 +30,10 @@ public ProjectEvaluationStartedEventArgs(string? message, params object[]? messa
         /// Gets or sets the full path of the project that started evaluation.
         /// </summary>
         public string? ProjectFile { get; set; }
+
+        /// <summary>
+        /// Gets or sets is the project is currently on restore phase.
+        /// </summary>
+        public bool IsRestore { get; internal set; }
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Framework/ProjectEvaluationStartedEventArgs.cs:37]: Actually - assuming that we also want to use this flag when replaying a binlog, we need to make a public surface change anyway. In other words, the new flag should likely be serialized. cc @JanKrivanek to confirm.

[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Framework/ProjectEvaluationStartedEventArgs.cs:37]: FYI @KirillOsenkov

[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Framework/ProjectEvaluationStartedEventArgs.cs:37]: Perhaps @JanKrivanek we should have a wiki page somewhere with a walkthrough of how to do BinaryLogger related changes? Including the viewer PR, version increments etc. Feel free to add to the wiki on the viewer repo or here, wherever it makes more sense for the team.

**Comment by JanKrivanek** [comment]
** UPDATE May/15 **
As decided collectively on a sync:
 * Let's try to see if we can put the GlobalVariables to `BuildStartedEventArgs`
 * Verify this covers tha case of `msbuild /t:restore`
 * This will probably cause that case of `masbuild /t:restore;build` will completely skip analyzers for the build part - this is accepted and fine. Let's just verify the fact and if true let's put it in the BuildCheck doc, plus let's create item for an analyzer that detects and flags this anti-pattern ('/t:restore;<another-target>')
 * The way of skipping the analysis during the resote (boolean flag, vs null proxy) can be discussed later on

File: src/Framework/ProjectEvaluationStartedEventArgs.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Framework/ProjectEvaluationStartedEventArgs.cs:37]: Good idea - I'll find a time to do this
